examples/postgres/custom_arrays/README.md

```diff
@@ -0,0 +1,892 @@
+will show up in the generates schema. Furthermore, you will need a wrapper struct to map from Rust to the Postgres type.
+For the ServiceEndpoint, Diesel already generated a matching wrapper struct with the correct annotations. The service
+table then uses that wrapper struct “ServiceEndpoint” instead of the native Postgres type to reference the custom type
+in the endpoints array.
+
+You want to attach the generated schema.rs to your lib file to access it from within your crate and stop your IDE from
+showing related warnings
+
+Next, you want to create a folder model with a mod file in it attached to your lib file. This is your place to store all
+Rust type implementations matching all the generated Postgres types and tables.
+
+Lastly, you also want to add a type alias for a pooled connection in the lib file because that connection definition is
+a long and convoluted type in need of a shorthand.
+
+At this point, your lib file looks as shown below:
+
+```rust
+mod schema;
+pub mod model;
+
+pub type Connection = diesel::r2d2::PooledConnection<diesel::r2d2::ConnectionManager<diesel::pg::PgConnection>>;
+```
+
+## Rust Types
+
+In total, you implement three Rust types:
+
+* Enum: ProtocolType
+* Struct: Endpoint
+* Struct: Service
+
+### Rust Enum: ProtocolType
+
+Diesel needs a wrapper struct to store a custom Enum in Postgres, so let’s add one:
+
+```rust
+#[derive(SqlType)]
+#[diesel(sql_type = protocol_type)]
+#[diesel(postgres_type(name = "protocol_type"))]
+pub struct PgProtocolType;
+```
+
+It is important that you add both, sql_type and postgres_type, otherwise insert will fail.
+
+Next, let’s define the actual Enum. Bear in mind to use the wrapper struct as sql_type on the Enum:
+
+```rust
+#[derive(Debug, Clone, FromSqlRow, AsExpression, PartialEq, Eq)]
+#[diesel(sql_type = PgProtocolType)]
+pub enum ProtocolType {
+    UnknownProtocol,
+    GRPC,
+    HTTP,
+    UDP,
+}
+```
+
+It’s worth pointing out that you are dealing with three types in total:
+
+1) protocol_type: The Postgres Enum type
+2) PGProtocolType: The Wrapper struct that refers to the Postgres Enum type
+3) ProtocolType: The Rust Enum that refers to the wrapper struct PGProtocolType
+
+Mixing any of those three types will result in a complicated Diesel trait error. However, these error messages are just
+a convoluted way to say that the database type mismatches the Rust type. When you encounter a consulted trait error
+message, make sure to check:
+
+1) Do I have a wrapper struct?
+2) Does my a wrapper struct derives SqlType?
+3) Does my wrapper type has both, sql_type and postgres_type declared?
+4) Are sql_type and postgres_type both referring to the correct Postgres type?
+5) Does my Rust type refers to the wrapper struct type?
+
+If all those checks pass and you still see errors, it’s most likely a serialization error.
+
+To serialize and deserialize a custom Enum, you write a custom toSql and fromSql implementation. Luckily, this is
+straightforward.
+
```

Comment:
I personally would like to add a paragraph here that explains what these impl do.

`ToSql` describes how to serialize a given rust type (`Self`) as sql side type (first generic argument) for a specific database backend (second generic argument). It needs to translate the type into the relevant wire protocol in that case. For postgres/mysql enums that's just the enum value as ascii string, but in general it's depended on the database + type.

`FromSql` describes how to deserialize a given rust type (`Self`) as sql side type (first generic argument) for a specific database backend (second generic argument). It need to translate from the relevant wire protocol to the rust type. For postgres/mysql enums that just means matching on the as ascii string, but in general it's depended on the database + type.

Reply:
That is a great explanation. I added it as it is to the Custom enum paragraph.